File: aoc/2024/aoc202011.hs
module AOC202011 where

import Data.Map(Map,findWithDefault,keys,toList)
import Data.Set(Set,empty,fromList,member,size)

import AOC

aoc = AOC {
    day="../../2020/input/11",
    aocTests=[
        AOCTest {
            testData=unlines [
                "L.LL.LL.LL",
                "LLLLLLL.LL",
                "L.L.L..L..",
                "LLLL.LL.LL",
                "L.LL.LL.LL",
                "L.LLLLL.LL",
                "..L.L.....",
                "LLLLLLLLLL",
                "L.LLLLLL.L",
                "L.LLLLL.LL"
                ],
            testResult=Just "37",
            testResult2=Just "26"
            }
        ],
    aocCode=Code {
        codeParse=parse neighbors1,
        codeParse2=parse neighbors2,
        codeTest=result 4,
        codeTest2=result 5,
        codeResult=result 4,
        codeResult2=result 5
        }
    }

parse neighbors input =
    map (neighbors xmax ymax grid) $ map fst $ filter ((== 'L') . snd)
                                             $ toList grid
  where
    grid = parse2d input
    xmax = maximum $ map fst $ keys grid
    ymax = maximum $ map snd $ keys grid

occupy :: Int -> Set (Int,Int) -> ((Int,Int),[(Int,Int)]) -> Bool
occupy threshold set (xy@(x,y),xys)
  | count == 0 = True
  | count >= threshold = False
  | otherwise = member xy set
  where
    count = length [() | xy <- xys, member xy set]

next :: Int -> [((Int,Int),[(Int,Int)])] -> Set (Int,Int) -> Set (Int,Int)
next threshold seats occupied =
    fromList $ map fst $ filter (occupy threshold occupied) seats

findFixedPoint :: Eq a => [Set a] -> Int
findFixedPoint (a1:rest@(a2:_))
  | a1 == a2 = size a1
  | otherwise = findFixedPoint rest

neighbors1 :: Int -> Int -> Map (Int,Int) Char -> (Int,Int)
           -> ((Int,Int),[(Int,Int)])
neighbors1 _ _ grid xy@(x,y) =
    (xy,[(x+dx,y+dy) | dx <- [-1,0,1], dy <- [-1,0,1],
                       (dx,dy) /= (0,0),
                       findWithDefault '.' (x+dx,y+dy) grid == 'L'])

neighbors2 :: Int -> Int -> Map (Int,Int) Char -> (Int,Int)
           -> ((Int,Int),[(Int,Int)])
neighbors2 xmax ymax grid xy@(x,y) =
    (xy, take 1 [xy | xy <- zip (repeat x) yplus, seat xy]
      ++ take 1 [xy | xy <- zip (repeat x) yminus, seat xy]
      ++ take 1 [xy | xy <- zip xplus (repeat y), seat xy]
      ++ take 1 [xy | xy <- zip xplus yplus, seat xy]
      ++ take 1 [xy | xy <- zip xplus yminus, seat xy]
      ++ take 1 [xy | xy <- zip xminus (repeat y), seat xy]
      ++ take 1 [xy | xy <- zip xminus yplus, seat xy]
      ++ take 1 [xy | xy <- zip xminus yminus, seat xy])
  where
    seat xy = findWithDefault '.' xy grid == 'L'
    xplus = [x+1,x+2..xmax]
    xminus = [x-1,x-2..0]
    yplus = [y+1,y+2..ymax]
    yminus = [y-1,y-2..0]

result threshold seats = findFixedPoint $ iterate (next threshold seats) empty

File: aoc/2025/aoc202011.go
package main

func init() {
	Register(&aoc202011{
		AOC: AOC{
			Day:           11,
			InputFilename: "../2020/input/11.txt",
			Tests: []Test{
				Test{
					Input: `L.LL.LL.LL
LLLLLLL.LL
L.L.L..L..
LLLL.LL.LL
L.LL.LL.LL
L.LLLLL.LL
..L.L.....
LLLLLLLLLL
L.LLLLLL.L
L.LLLLL.LL
`,
					Part1: "37",
					Part2: "26",
				},
			},
		},
	})
}

type aoc202011 struct {
	AOC
}

func (aoc *aoc202011) parse(part1 bool, input *Input) [][][2]int {
	w, h, grid := input.Grid()
	seats := [][][2]int{}
	for xy, ch := range grid {
		if ch != 'L' {
			continue
		}
		seat := [][2]int{xy}
		if part1 {
			for _, n := range [...]XY{
				AdvanceXY(xy, DirR, 1),
				AdvanceXY(xy, DirR|DirD, 1),
				AdvanceXY(xy, DirD, 1),
				AdvanceXY(xy, DirD|DirL, 1),
				AdvanceXY(xy, DirL, 1),
				AdvanceXY(xy, DirL|DirU, 1),
				AdvanceXY(xy, DirU, 1),
				AdvanceXY(xy, DirU|DirR, 1),
			} {
				if grid[n] == 'L' {
					seat = append(seat, n)
				}
			}
		} else {
			for _, dir := range [...]int{DirR, DirR | DirD, DirD, DirD | DirL, DirL, DirL | DirU, DirU, DirU | DirR} {
				n := AdvanceXY(xy, dir, 1)
				for n[0] >= 0 && n[0] < w && n[1] >= 0 && n[1] < h {
					if grid[n] == 'L' {
						seat = append(seat, n)
						break
					}
					n = AdvanceXY(n, dir, 1)
				}
			}
		}
		seats = append(seats, seat)
	}
	return seats
}

func (aoc *aoc202011) result(part1 bool, input *Input) string {
	threshold := 5
	if part1 {
		threshold = 4
	}
	seats := aoc.parse(part1, input)
	area := map[XY]bool{}
	tmp := map[XY]bool{}
	changed := true
	for changed {
		changed = false
		clear(tmp)
		for _, seat := range seats {
			count := 0
			for _, n := range seat[1:] {
				if area[n] {
					count++
				}
			}
			if area[seat[0]] {
				if count >= threshold {
					changed = true
				} else {
					tmp[seat[0]] = true
				}
			} else {
				if count == 0 {
					changed = true
					tmp[seat[0]] = true
				}
			}
		}
		area, tmp = tmp, area
	}
	return IntResult(len(area))
}

func (aoc *aoc202011) Part1(input *Input) string {
	return aoc.result(true, input)
}

func (aoc *aoc202011) Part2(input *Input) string {
	return aoc.result(false, input)
}
